opengsl/method/models/wsgnn.py

```diff
@@ -0,0 +1,244 @@
+from torch_sparse import SparseTensor
+from torch_geometric.nn.conv.gcn_conv import gcn_norm
+import torch.nn as nn
+import torch.nn.functional as F
+import torch
+from torch_geometric.nn import GCNConv, GATConv, APPNP
+import torch_sparse
+
+
+class GraphLearner(nn.Module):
+    def __init__(self, input_size, num_pers=16):
+        super(GraphLearner, self).__init__()
+        self.weight_tensor = torch.Tensor(num_pers, input_size)
+        self.weight_tensor = nn.Parameter(nn.init.xavier_uniform_(self.weight_tensor))
+
+    def reset_parameters(self):
+        self.weight_tensor = nn.Parameter(nn.init.xavier_uniform_(self.weight_tensor))
+
+    def forward(self, context):
+        expand_weight_tensor = self.weight_tensor.unsqueeze(1)
+        context_fc = context.unsqueeze(0) * expand_weight_tensor
+        context_norm = F.normalize(context_fc, p=2, dim=-1)
+        attention = torch.matmul(context_norm, context_norm.transpose(-1, -2)).mean(0)
+        mask = (attention > 0).detach().float()
+        attention = attention * mask + 0 * (1 - mask)
+
+        return attention
+
+class MLP(nn.Module):
+    """ adapted from https://github.com/CUAI/CorrectAndSmooth/blob/master/gen_models.py """
+    def __init__(self, in_channels, hidden_channels, out_channels, num_layers,
+                 dropout=.5, use_bn=False):
+        super(MLP, self).__init__()
+        self.use_bn = use_bn
+        self.lins = nn.ModuleList()
+        self.bns = nn.ModuleList()
+        if num_layers == 1:
+            # just linear layer i.e. logistic regression
+            self.lins.append(nn.Linear(in_channels, out_channels))
+        else:
+            self.lins.append(nn.Linear(in_channels, hidden_channels))
+            self.bns.append(nn.BatchNorm1d(hidden_channels))
+            for _ in range(num_layers - 2):
+                self.lins.append(nn.Linear(hidden_channels, hidden_channels))
+                self.bns.append(nn.BatchNorm1d(hidden_channels))
+            self.lins.append(nn.Linear(hidden_channels, out_channels))
+
+        self.dropout = dropout
+
+    def reset_parameters(self):
+        for lin in self.lins:
+            lin.reset_parameters()
+        for bn in self.bns:
+            bn.reset_parameters()
+
+    def forward(self, x):
+        for i, lin in enumerate(self.lins[:-1]):
+            x = lin(x)
+            x = F.relu(x, inplace=True)
+            if self.use_bn:
+                x = self.bns[i](x)
+            x = F.dropout(x, p=self.dropout, training=self.training)
+        x = self.lins[-1](x)
+        return x
+
+
+class DenseAPPNP(nn.Module):
```

Comment:
Can you replace this part with our implemented APPNP?

Reply:
It is replaced now